[RDY] More fixes to the new TUI

tarruda opened the pull request:
Fix #2005 #2016 #2017 #2024

tarruda committed:
tui: Only send the pastetoggle key if necessary

renstrom commented:
Verified.

tarruda committed:
tui: Replace uv_tty_t by uv_pipe_t

It seems uv_tty_t doesn't work properly on OSX(it is set to blocking mode,
which causes UV_RUN_NOWAIT to misbehave).

oni-link commented on src/nvim/tui/term_input.inl:
```diff
@@ -259,8 +265,8 @@ static TermInput *term_input_new(void)
-  uv_tty_init(uv_default_loop(), &rv->input_handle, rv->in_fd, 1);
-  uv_tty_set_mode(&rv->input_handle, UV_TTY_MODE_RAW);
+  uv_pipe_init(uv_default_loop(), &rv->input_handle, 0);
+  uv_pipe_open(&rv->input_handle, 0);
```
Is `in_fd` not used anymore?

tarruda replied:
forgot about it :+1:

tarruda committed:
tui: Fix terminfo extension indexes

For unibilium extension indexes, use signed integer type initialized with -1 to
distinguish from the first extension string which will always have index 0.

tarruda committed:
tui: Set default color attributes when scrolling

This is necessary for some terminals, or else they will clear/insert lines with
wrong background color

tarruda committed:
tui: Don't use 'timeout' for determining esc timeout

tarruda commented:
merging after tests
